### DISCUSSION TITLE
recorded_value stalls on out of scan tag

### DISCUSSION THREAD

**Discussion by IndyAbarth** [Hugovdberg/PIconnect#802] [Q&A]
I have a python program to select tags and pull data at user specified intervals using recorded_value. It seems that if the user selects an out-of-scan tag, then recorded_value stalls and the program must be shutdown. I can't find any way to filter those tags out of the user selectable list to prevent this from happening. Is there a way to accomplish this??

**Comment by Hugovdberg** [comment]
I don't think this is currently possible. I also don't know why the program would stall when it is selected, do you look up some data when the user selects the tag or do you mean the call to recorded_values stalls? You could try to access the PIpoints in a separate thread from your user interface and then set a timeout, after which you kill the thread and report the error to the user.

**Reply by IndyAbarth** [reply]
I have the user select the tags in a tkinter form and set the date/time to start and stop, and the interval. On certain tags, not just out-of-scan tags, the recorded_value call stalls out. Trimmed down to a basic form, this is what stalls:

```
from PIConnect.PI import *
import datetime
import pandas as pd
import time

with PIServer('cinstpi') as server:
    dt=datetime.datetime.now()
    print(dt)
    tag_name="GIBSON CIR 34507 PH A"
    tag_val=list(server.search(tag_name, point_list=True).recorded_value(dt).values())[-1][-1]
    print(tag_name + " ", tag_val)
```

I get the output of "Print(dt)" but then execution stalls and I have to close VS Code. I know this particular tag has good values because I can trend it in PI Vision. This will work fine on other tags. 

I'm still new to Python but I experimented with threading based on internet searches, and I found that I couldn't get anything to run concurrent with the recorded_value call.

**Reply by Hugovdberg** [reply]
Does this code also stall when you run it directly, or just when it is called in your tkinter app? Graphical interfaces are notorious for stalls when a job is running in the same thread/event loop as the interface itself. 
Could you separate the server.search call from the recorded_values call? I'm curious in which of those the stall occurs.

**Reply by IndyAbarth** [reply]
what I posted was a trimmed down version that does stall when run in VS Code. I separated everything out: 
```
from PIConnect.PI import *
import datetime
import pandas as pd
import time

with PIServer('cinstpi') as server:
    dt=datetime.datetime.now()
    print(dt)
    tag_name="GIBSON CIR 34507 PH A"
    point=server.search(tag_name, point_list=True)
    print(point)
    rc_val=point.recorded_value(dt)
    print(rc_val)
    tag_val=list(rc_val.values())[-1][-1]
    print(tag_val)
```

When I run this with a working tag, I get info in the terminal as expected (actual data replaced with '***'):
```
2025-01-18 11:14:31.898680
PIPointList(Contains 1 PIPoints)
{'GIBSON CIR 34507 MVA': (Timestamp('2025-01-18 16:14:31.898681+0000', tz='UTC'), ************)}
************
```

When I do it with the "bad" tag, it actually appears to stall at the search call, because the terminal only shows the datetime:
`2025-01-18 11:14:46.062798`

This is especially interesting because, in the main program, I can search for 'gibson', and the "bad" tag shows up in the list with no delay.

**Reply by IndyAbarth** [reply]
I never figured out why this is happening, but I was able to find a better method of working with the PIPoints to circumvent the issue.
